modules/scripts-utils/ace-adapter/src/main/java/org/daisy/pipeline/epub/AceProvider.java

```diff
@@ -0,0 +1,181 @@
+
+		@Override
+		public void setOutput(String port, WritablePipe pipe) {
+			if ("json-report".equals(port)) {
+				jsonReportData = pipe;
+			} else { // default output to htmlReportURI
+				htmlReportURIport = pipe;
+			}
+		}
+
+		@Override
+		public void reset() {
+			htmlReportURIport.resetWriter();
+			jsonReportData.resetWriter();
+		}
+
+		@Override
+		public void run() throws SaxonApiException {
+			super.run();
+			try {
+				URI epubURI = new URI(getOption(_epubFile).getString());
+				File epubFile = new File(epubURI);
+				
+				
+				if(aceProgram == null) throw new Exception("(running pxi:ace) ACE was not found in the PATH of your system");
+
+				String path = epubFile.getCanonicalPath();
+				
+				// Output where the ace reports (report.html and report.json) and unzipped epub will be stored
+				File tempDir;
+				if(getOption(_tempDir).getString().equals("")) {
+					tempDir = Files.createTempDirectory("ace-").toFile();
+				}else tempDir = new File(new URI(getOption(_tempDir).getString())); 
+				
+				System.out.println(getOption(_tempDir).getString());
+				System.out.println(tempDir.getAbsolutePath());
+				
+				String language = getOption(_lang).getString();
+				
+				String[] cmd = new String[] {
+						aceProgram.getAbsolutePath(),
+						"-o", tempDir.getAbsolutePath(),
+						"-t", tempDir.getAbsolutePath(),
+						"-l", language.equals("") ? "en" : language,
+						path
+				};
+				
+				// NOTE : the command runner would raised an IOException on windows if the consumeOutput is not used
+				// (the CommandRunner pipes the process output to the file '/dev/null' by default)
+				new CommandRunner(cmd)
+						.consumeOutput(stream -> {})
+						.consumeError(mLogger)
+						.run();
+				
+				
+				File htmlReport = new File(tempDir.getAbsolutePath() + File.separator + "report.html");
+				File jsonReport = new File(tempDir.getAbsolutePath() + File.separator + "report.json");
+				
+				// write the results uri in temps document, and pipe them to the output ports as xml documents
+				// (reading the html report does not work here : the ace report is not strictly valid, 
+				// some non closed link tags are blocking the load by saxon
+				File tempFile = File.createTempFile("html-uri-",".xml");
+				PrintWriter tempWriter = new PrintWriter(tempFile,"UTF-8");
+				tempWriter.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
+				tempWriter.println("<c:result xmlns:c=\"http://www.w3.org/ns/xproc-step\">" + htmlReport.toURI().toString() + "</c:result>");
+				tempWriter.close();
+				htmlReportURIport.write(runtime.getProcessor().newDocumentBuilder().build(tempFile));
+				tempFile.delete();
+
+				String jsonContent = new String(Files.readAllBytes(Paths.get(jsonReport.getAbsolutePath())),StandardCharsets.UTF_8)
+						.replaceAll("<", "&#x3e;")
+						.replaceAll(">", "&#x3c;")
+						.replaceAll("&", "&#x26;");
+				
+				if (!jsonContent.isBlank()) {
+					tempFile = File.createTempFile("json-data",".xml");
```

Comment:
I would put this code for converting the JSON file to a `c:data` XML document in a separate function. It could even be made into a utility function in `file-utils` for instance, and it doesn't need to be specific to JSON (you could pass the content-type as argument). There could also be an XProc variant of this Java function, based on the following step internally declared inside `px:fileset-load`:

```xml
<p:declare-step type="pxi:load-text">
    <p:output port="result"/>
    <p:option name="href"/>
    <p:identity>
      <p:input port="source">
        <p:inline>
          <c:request method="GET" override-content-type="text/plain; charset=utf-8"/>
        </p:inline>
      </p:input>
    </p:identity>
    <p:add-attribute match="c:request" attribute-name="href">
      <p:with-option name="attribute-value" select="$href"/>
    </p:add-attribute>
    <p:http-request/>
  </p:declare-step>
```

Or it could use the Java implementation of course.

Maybe the code above for creating the `c:result` document could even be made into a common utility function. This is being done in several other steps so this is a way to simplify all that code.

I can do this myself.